### PR TITLE
Import changes from the overlay for Data mutation of slices

### PR DESCRIPTION
**Pull request by phausler**
Mutation of Data slices had issues when it comes to re-indexing to the base slice region. Previously when a slice of Data was mutated the re-calculation of the slice region was incorrectly assigned to a relative region to the previous backing. Now the slice region will copy the specific region out creating a new data upon CoW semantics and offset the index region upon access. Mutating a non CoW slice will retain the previous region backing but adjust the remaining allocation buffer accordingly. To validate this I have added tests for combinations that should approximate all possible combinations of mutations, backing storages and CoW scenarios.

This fixes the following issues:
rdar://problem/34206043
SR-5887
SR-5873
SR-5810

tl;dr – Slicing a CoW should result in fillet mignon not ground chuck.

NOTE: Data has a failure set for custom backing mutation because NSData does not properly delegate to CFData because of missing factory pattern initializers. This prevents testing of a number of mutation cases and should be re-enabled once we have a suitable solution for factory initializers.

### CODE REVIEW

**Comment by spevans** [comment]
@phausler you may want to use ` _CFIsMainThread()` instead of `pthread_main_np()`

**Comment by phausler** [comment]
hmm good point

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test
